lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,350 @@
+///   are otherwise valid field names, we prepend the underscore.
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer {
+    valid_regex: Regex,
+    conformed_log: Option<LogEvent>,
+}
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer {
+            valid_regex: Regex::new(r"^[\w\.\-]*$").unwrap(),
+            conformed_log: None,
+        }
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        match event {
+            Event::Log(log) => serde_json::to_value(&log),
+            // TODO Metric and Trace shouldn't be valid but I was not finding a way to construct a
+            // custom serde_json::Error on the fly
+            Event::Metric(metric) => serde_json::to_value(&metric),
+            Event::Trace(trace) => serde_json::to_value(&trace),
+        }
+    }
+
+    /// Validates that the GELF required fields exist in the event.
+    fn validate_required_fields(&mut self, log: &LogEvent) -> vector_core::Result<()> {
+        if !log.contains(VERSION) {
+            return Err(format!("LogEvent does not contain required field '{}'", VERSION).into());
+        }
+        if !log.contains(HOST) {
+            return Err(format!("LogEvent does not contain required field '{}'", HOST).into());
+        }
+
+        let message_key = log_schema().message_key();
+        if log.contains(message_key) {
+            self.conformed_log = Some(log.clone());
+            self.conformed_log
+                .as_mut()
+                .unwrap()
+                .rename_key(message_key, SHORT_MESSAGE);
+        } else if !log.contains(SHORT_MESSAGE) {
+            return Err(format!(
+                "LogEvent does not contain required field '{}'",
+                SHORT_MESSAGE
+            )
+            .into());
+        }
```

Comment:
This reads awkwardly to me, I suspect there's a better way of modeling it...

Reply:
took a stab, lmk if its not better